Support DAG Versioning on the grid view

jedcunningham opened the pull request:
#48534 needs to be merge first.

Time to support DAG Versioning on the grid view. At a high level, this means the grid view will handle situations changing DAG structure between runs. For example, if you remove a task, the grid view will now continue to show that task for existing runs that had it.

In order to accomplish this, we have a new combined structure in the grid view api response - the mash up of all of the runs in the DAG.

There are still some corner cases that aren't rendered well (or at all?), but this covers the normal "removed a task" or "removed a task group" scenarios well.

Here is an example:
<img width="322" alt="Screenshot 2025-03-27 at 11 33 15 PM" src="https://github.com/user-attachments/assets/5a00dc85-cc33-4f0f-903a-7b2614deec45" />

`real_end` was added for the second run only, and `inner_section_2` was also removed for the 3rd run.

Closes: #47857

jedcunningham committed:
Support DAG Versioning on the grid view

Time to support DAG Versioning on the grid view. At a high level, this
means the grid view will handle situations changing DAG structure
between runs. For example, if you remove a task, the grid view will now
continue to show that task for existing runs that had it.

In order to accomplish this, we have a new combined structure in the
grid view api response - the mash up of all of the runs in the DAG.

There are still some corner cases that aren't rendered well (or at
all?), but this covers the normal "removed a task" or "removed a task
group" scenarios well.

Co-authored-by: Brent Bovenzi <brent@astronomer.io>

phanikumv reviewed (approved):
Amazing work on this !

bugraoz93 reviewed (approved):
Looks great!

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/ui/grid.py:
```diff
@@ -146,33 +153,36 @@ def grid_data(
-        elif parent_id and task_node_map[parent_id]["is_group"]:
-            parent_tis[(parent_id, ti.run_id)].append(ti)
+
+    for _, tis in tis_by_run_id.items():
```
Nit
```suggestion
    for tis in tis_by_run_id.values():
```

jedcunningham replied:
Thanks, I swear I had done this at one point...

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/services/ui/grid.py:
```diff
@@ -237,3 +243,41 @@ def fill_task_instance_summaries(
+        if node["id"] in current_ids:
+            current_node = _get_node_by_id(current, node["id"])
+            # if we have children, merge those as well
+            if "children" in current_node:
+                _merge_node_dicts(current_node["children"], node["children"])
```
Nice, that's the part I was looking for

pierrejeambrun reviewed (approved):
Great 🎉 thanks

jedcunningham committed:
Update airflow-core/src/airflow/api_fastapi/core_api/routes/ui/grid.py

Co-authored-by: Pierre Jeambrun <pierrejbrun@gmail.com>

jedcunningham committed:
Restore "serialize once" behavior for a couple asset tests

I'm not convinced these tests and/or dag_maker are actually right...
but this is how they were.

jedcunningham commented on airflow-core/tests/unit/models/test_taskinstance.py:
```diff
@@ -4122,7 +4122,7 @@ def first_asset_task(*, outlet_events):
             first_asset_task()
 
-        with dag_maker(schedule=None, serialized=True, session=session):
+        with dag_maker(schedule=None, serialized=False, session=session):
```
@uranusjr @Lee-W pretty sure this is a sign these 2 TI tests are either wrong and/or there is a gremlin hiding in dag_maker around assets?

Lee-W replied:
I think serialized is needed for running dag with asset. Looks like an asset won't be activated if it's not serialized and a dag cannot be ran with an inactive asset.

https://github.com/apache/airflow/blob/075d0dd304d6f036ad180b07dec5e808e33d78fc/devel-common/src/tests_common/pytest_plugin.py#L777

jedcunningham replied:
As part of this PR, I fixed dag_maker to create more than 1 serdag record. So these 2 tests will now try to create 2 serdag records, but during the second one, I get this:

```
E   sqlalchemy.exc.IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "idx_asset_active_name_unique"
E   DETAIL:  Key (name)=(asset_first) already exists.
E
E   [SQL: INSERT INTO asset_active (name, uri) VALUES (%(name)s, %(uri)s)]
E   [parameters: {'name': 'asset_first', 'uri': 'test://asset/'}]
```

And, if I turn off serialization for the second one (restore the old behavior), it passes.

I don't really understand these tests either - how can these 2 assets be inactive if they are in the latest parse of the DAG?

I'm temped to put a skip on them, thoughts?

jedcunningham replied:
Okay, so shoutout to @Lee-W for having a quick pairing session - we fixed 1 and decided to skip the second for now. #48539 has been opened to track getting it back running.

jedcunningham committed:
Sort children when deserializing DAG

MySQL doesn't seem to maintain the order of children round trip.

jedcunningham committed:
Remove `version_number` from grid response

jedcunningham committed:
Fix 1 TI test, skip another - also deactivate assets in dag_maker

jedcunningham commented on airflow-core/src/airflow/serialization/serialized_objects.py:
```diff
@@ -1865,7 +1865,7 @@ def set_ref(task: Operator) -> Operator:
                 else cls.deserialize_task_group(val, group, task_dict, dag=dag)
             )
-            for label, (_type, val) in encoded_group["children"].items()
+            for label, (_type, val) in sorted(encoded_group["children"].items(), key=lambda x: x[0])
```
```suggestion
            for label, (_type, val) in sorted(encoded_group["children"].items())
```

#48534 will do this change first, with a test, but to fix test here I'm going to leave this in here.

jedcunningham committed:
Skip both breaking asset tests

pierrejeambrun reviewed (approved):
Looking good :)

jedcunningham committed:
Merge branch 'main' into abandoned-tasks-grid
